### ISSUE TITLE
Ultra small issue In daemon.c with realpath() call

### ISSUE THREAD

**Issue by matteotenca** [pkoutoupis/rapiddisk#130]
Hi,

using valgrind and compiling rapiddiskd with `-Og -g -Wl,--export-dynamic -gdwarf` options, I found out that `rapiddiskd` can't start:

```bash
shub@gturismo:/tmp/rapiddisk/src$ sudo ./rapiddiskd -h
rapiddiskd 8.2.0
Copyright 2011 - 2022 Petros Koutoupis

*** buffer overflow detected ***: terminated
Aborted
shub@gturismo:/tmp/rapiddisk/src$
```

So, I passed it to valgrind:

```
shub@gturismo:/tmp/rapiddisk/src$ make && sudo valgrind --sim-hints=lax-ioctls --show-leak-kinds=all --leak-check=full --track-origins=yes -s ../src/rapiddiskd -h
Make has completed.
==27528== Memcheck, a memory error detector
==27528== Copyright (C) 2002-2017, and GNU GPL'd, by Julian Seward et al.
==27528== Using Valgrind-3.15.0 and LibVEX; rerun with -h for copyright info
==27528== Command: ../src/rapiddiskd -h
==27528==
rapiddiskd 8.2.0
Copyright 2011 - 2022 Petros Koutoupis

*** buffer overflow detected ***: terminated
==27528==
==27528== Process terminating with default action of signal 6 (SIGABRT)
==27528==    at 0x490000B: raise (raise.c:51)
==27528==    by 0x48DF858: abort (abort.c:79)
==27528==    by 0x494A26D: __libc_message (libc_fatal.c:155)
==27528==    by 0x49ECAB9: __fortify_fail (fortify_fail.c:26)
==27528==    by 0x49EB355: __chk_fail (chk_fail.c:28)
==27528==    by 0x49EB927: __realpath_chk (realpath_chk.c:29)
==27528==    by 0x10AA68: realpath (stdlib.h:45)
==27528==    by 0x10AA68: main (daemon.c:152)
==27528==
==27528== HEAP SUMMARY:
==27528==     in use at exit: 96,210 bytes in 694 blocks
==27528==   total heap usage: 1,858 allocs, 1,164 frees, 440,083 bytes allocated
==27528==

[...]

Aborted
shub@gturismo:/tmp/rapiddisk/src$
```

The pieces of code involved are:

https://github.com/pkoutoupis/rapiddisk/blob/2e1f8a18a61bd31ca61489c38cf139746d25a2ca/src/daemon.c#L152-L156

and:

https://github.com/pkoutoupis/rapiddisk/blob/2e1f8a18a61bd31ca61489c38cf139746d25a2ca/src/daemon.c#L122

It was strange, but after some digging I found that `__realpath_chk()` (docs [here](http://refspecs.linux-foundation.org/LSB_4.0.0/LSB-Core-generic/LSB-Core-generic/libc---realpath-chk-1.html)) exits with an error if the size of the buffer provided (the size of `path`) is < then `PATH_MAX` as defined in `limits.h`.

Long story short, the best behaviour is to call `realpath` in such a way it takes care of that buffer:

```C
char *path;

[...]

if ((path = realpath(argv[0], NULL)) == NULL) {
[...]

if (path) free(path);
```

This way, realpath malloc's the needed memory; then we have to `free()` it later on.
This approach is detailed [here](https://man7.org/linux/man-pages/man3/realpath.3.html#BUGS).

All in all, it is a minimal issue (or not even an issue), but I found it interesting to point it out. I found it just because I was using `valgrind` here and there looking for memoey leaks.

Regards

**Comment by pkoutoupis** [comment]
I have got a big change that I want to merge tomorrow (or maybe Sunday): https://github.com/pkoutoupis/rapiddisk/pull/123. Once I merge that, I will address this ticket with a couple of low lying adjustments in another PR which I will add you and then I will dive back to the first and much larger set of memory leaks in your PR. As always, thank you so much for your help! :-D

**Comment by matteotenca** [comment]
You're welcome!
